Add initial implementation of VFF-based outdoor navigation using GPS data

jmguerreroh opened the pull request:
Hey! The robot is moving outside! 🎉 Play for fun and test the new VFF-based GPS navigation module. 
Feedback and improvements are welcome!

jmguerreroh committed:
Initial implementation of VFF-based outdoor navigation using GPS data

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Initial implementation of VFF-based outdoor navigation using GPS data

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Initial implementation of VFF-based outdoor navigation using GPS data

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Initial implementation of VFF-based outdoor navigation using GPS data

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+  node->declare_parameter<float>(plugin_name + ".obstacle_detection_y_max", 10.0);
+  node->declare_parameter<float>(plugin_name + ".obstacle_detection_z_min", 0.10);
+  node->declare_parameter<float>(plugin_name + ".obstacle_detection_z_max", 1.00);
+  node->declare_parameter<double>(plugin_name + ".max_speed", 0.8);
```
Maybe, in a future improved version, we should also incorporate the acceleration, and `max_speed` will be reached by increasing the current velocity using the time and the acceleration.

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+
+  // Create subscriber to IMU data
+  imu_subscriber_ = node->create_subscription<sensor_msgs::msg::Imu>(
+    "/imu/data", rclcpp::SensorDataQoS().reliable(),
```
`"imu/data"` to allow namespaces

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+    std::bind(&GpsVffController::imu_callback, this, std::placeholders::_1));
+
+  // Publisher for visualization markers
+  marker_array_pub_ = node->create_publisher<visualization_msgs::msg::MarkerArray>(
```
`"markers_vff` (or even `"vff/markers_vff`)

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+  return {};
+}
+
+void GpsVffController::imu_callback(const sensor_msgs::msg::Imu::SharedPtr msg)
+{
+  // Extract the yaw angle from the IMU data
+  tf2::Quaternion q(
+    msg->orientation.x, msg->orientation.y,
+    msg->orientation.z, msg->orientation.w);
+  tf2::Matrix3x3 m(q);
+  double roll, pitch, yaw;
+  m.getRPY(roll, pitch, yaw);
+  yaw_ = yaw;
+}
```
Isn't this more related to the localization than the control? Maybe it should be moved to the localization package...

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+  vff_vector.result[0] = vff_vector.attractive[0] + vff_vector.repulsive[0];
+  vff_vector.result[1] = vff_vector.attractive[1] + vff_vector.repulsive[1];
+
+  // Publish debug markers
+  visualization_msgs::msg::Marker marker;
+  marker.header.frame_id = cmd_vel_.header.frame_id;
+  marker.header.stamp = get_node()->now();
+  marker.type = visualization_msgs::msg::Marker::SPHERE;
+  marker.id = 456;
+  marker.pose.position.x = closest_point.x;
+  marker.pose.position.y = closest_point.y;
+  marker.pose.position.z = closest_point.z;
+  marker.pose.orientation.w = 1.0;
+  marker.scale.x = marker.scale.y = marker.scale.z = 0.1;
+  marker.color.a = 1.0;
+  marker.color.r = 1.0;
+  marker.color.g = 0.0;
+  marker.color.b = 0.0;
+
+  auto marker_array = get_debug_vff(vff_vector, frame_id);
+  marker_array.markers.push_back(marker);
+
```
Maybe this should be inside the `if (marker_array_pub_->get_subscription_count() > 0) ` for avoiding extra computation if nobody is waiting for the markers.

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+  vff_vector.repulsive = {0.0, 0.0};
+  vff_vector.result = {0.0, 0.0};
+
+  // Find the closest obstacle in the point cloud
+  pcl::PointXYZ closest_point;
+  double min_distance = std::numeric_limits<double>::max();
+
+  for (const auto & point : pointcloud_) {
+    const double distance = std::hypot(point.x, point.y);
+    if (distance < min_distance) {
+      min_distance = distance;
+      closest_point = point;
+    }
+  }
```
We should consider all the points within a working radius in a future improved version.

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    // Get the fused PointCloud
+    auto fused =
```
We should take time measurements to ensure that this method takes less than ~10ms. 

I would start this pipeline with a filter to consider only the nearby points _a grosso modo_, and check it if this makes this line more efficient, as the `fuse` is the expensive step in this pipeline :

```
auto fused =
      PerceptionsOpsView(nav_state.perceptions)
            .filter({-10.0, -10.0, -10.0}, ({10.0, 10.0, 10.0})
            .fuse(cmd_vel_.header.frame_id)
            ->filter({obstacle_detection_x_min_, obstacle_detection_y_min_, obstacle_detection_z_min_{obstacle_detection_x_max_, obstacle_detection_y_max_,
        obstacle_detection_z_max_}).as_points();
```

fmrico commented on easynav_gps_vff/easynav_gps_vff_localizer/src/easynav_gps_vff_localizer/GpsVffLocalizer.cpp:
```diff
@@ -0,0 +1,103 @@
+  // Get XY cartesian coordinates respect to the origin
+  odom_.pose.pose.position.x = utm_x - origin_utm_.x;
+  odom_.pose.pose.position.y = utm_y - origin_utm_.y;
+}
```
I would incorporate here the orientation with the IMU

fmrico commented on easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp:
```diff
@@ -0,0 +1,306 @@
+  double current_y_ = odom.pose.pose.position.y;
+
+  // If a goal is set
+  if (!all_goals.goals.empty()) {
```
Maybe we should check here to see if there is a path. If no path, use the goals, but if the path is available, we should set as a local goal a point in the path to some short distance, like 1 meter. Another option is using parameters for changing between path/goal mode.

jmguerreroh replied:
I will work on this feature in a future PR

jmguerreroh committed:
Separated GPS and VFF logic

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Separated GPS and VFF logic

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Separated GPS and VFF logic

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh committed:
Update README.md

jmguerreroh commented:
Hey @fmrico! I really appreciate your fabulous comments.
I've done my best to implement them, and I hope these changes meet your expectations. 😊

jmguerreroh committed:
Remove old packages

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Remove old packages

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Remove old packages

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Remove old packages

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Update README.md & Sync upstream

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Update README.md & Sync upstream

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Update README.md

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Add new packages to CI

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Add new packages to CI

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

fmrico commented on easynav_gps_localizer/include/easynav_gps_localizer/GpsLocalizer.hpp:
```diff
@@ -0,0 +1,183 @@
+/// \file
+/// \brief Declaration of the GpsLocalizer class, a default plugin implementation for localization.
+
+#ifndef EASYNAV_LOCALIZER__GpsLOCALIZER_HPP_
```
EASYNAV_LOCALIZER__GPSLOCALIZER_HPP_

jmguerreroh replied:
Done

fmrico commented on easynav_gps_localizer/include/easynav_gps_localizer/GpsLocalizer.hpp:
```diff
@@ -0,0 +1,183 @@
+   *
+   * Stores the estimated position and velocity of the robot.
+   */
+  nav_msgs::msg::Odometry odom_ {};
```
The `{}` is not necessary

jmguerreroh replied:
I followed this example (:

https://github.com/EasyNavigation/EasyNavigation/blob/rolling/easynav_localizer/include/easynav_localizer/DummyLocalizer.hpp

fmrico commented on easynav_gps_localizer/include/easynav_gps_localizer/GpsLocalizer.hpp:
```diff
@@ -0,0 +1,183 @@
+   * Stores the first UTM coordinates obtained from the GPS data.
+   * It is used to calculate the relative position of the robot in the UTM coordinate system.
+   */
+  geometry_msgs::msg::Point origin_utm_ {};
```
The `{}` is not necessary

jmguerreroh replied:
I followed this example (:

https://github.com/EasyNavigation/EasyNavigation/blob/rolling/easynav_localizer/include/easynav_localizer/DummyLocalizer.hpp

fmrico commented on easynav_gps_localizer/include/easynav_gps_localizer/GpsLocalizer.hpp:
```diff
@@ -0,0 +1,183 @@
+   * It is used to broadcast the static transform between the "map" frame and the "odom" frame.
+   */
+  std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;
+
```
This plugin should also have a `tf2_ros::TransformBroadcaster` and publish the transform `map->odom`

jmguerreroh replied:
LGTM I will work on this feature in a future PR

fmrico commented on easynav_gps_localizer/src/easynav_gps_localizer/GpsLocalizer.cpp:
```diff
@@ -0,0 +1,120 @@
+
+void GpsLocalizer::update_rt(const NavState & nav_state)
+{
+
+
```
Remove these lines

jmguerreroh replied:
Done

fmrico commented on easynav_vff_controller/src/easynav_vff_controller/VffController.cpp:
```diff
@@ -0,0 +1,299 @@
+    double dy = goal_.y - current_y_;
+
+    // Calculate the Euclidean distance to the goal
+    double distance = std::hypot(dx, dy);  // more stable than sqrt(pow(..., 2))
```
Remove these ChatGPT-like comments ;)

`// more stable than sqrt(pow(..., 2))`

jmguerreroh replied:
Done :')

fmrico commented:
Hi @jmguerreroh 

Only a few comments. Maybe the more challenging part is publishing `map->odom`. We can move this to a future PR, as well as the path following.

jmguerreroh committed:
Minor bug fixes

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Minor bug fixes

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Merge branch 'rolling' of https://github.com/jmguerreroh/easynav_outdoor_stack into rolling

jmguerreroh committed:
Merge branch 'EasyNavigation:rolling' into rolling

jmguerreroh committed:
Merge branch 'rolling' of https://github.com/jmguerreroh/easynav_outdoor_stack into rolling

jmguerreroh committed:
Fix dependencies

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Fix: replace deprecated LLtoUTM with GeographicLib::UTMUPS

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh committed:
Fix: replace deprecated LLtoUTM with GeographicLib::UTMUPS

Signed-off-by: Jose Miguel <josemiguel.guerrero@urjc.es>

jmguerreroh commented:
Hi @fmrico,

I’ve implemented the suggestions you mentioned. What’s still pending is the `map -> odom` transform using a `tf2_ros::TransformBroadcaster`, and modifying the VFF to consider all points within a given radius.

I’ve also replaced the deprecated `LLtoUTM` with `GeographicLib::UTMUPS`.

CI is now working like a charm. :trophy: :fireworks: :tada:

fmrico commented:
Great work!! 🔥 

Some functionality for next iterations, but i think this ok 

Merging!! 🚀
